test/e2e/pkg/grammar/checker.go

```diff
@@ -0,0 +1,212 @@
+// This method needs to be modified if we add another ABCI call.
+func (g *GrammarChecker) isSupportedByGrammar(req *abci.Request) bool {
+	switch req.Value.(type) {
+	case *abci.Request_InitChain, *abci.Request_FinalizeBlock, *abci.Request_Commit,
+		*abci.Request_OfferSnapshot, *abci.Request_ApplySnapshotChunk, *abci.Request_PrepareProposal,
+		*abci.Request_ProcessProposal:
+		return true
+	default:
+		return false
+	}
+}
+
+// filterRequests returns requests supported by grammar and remove the last height.
+func (g *GrammarChecker) filterRequests(reqs []*abci.Request) []*abci.Request {
+	var r []*abci.Request
+	for _, req := range reqs {
+		if g.isSupportedByGrammar(req) {
+			r = append(r, req)
+		}
+	}
+	r, _ = g.filterLastHeight(r)
+	return r
+}
+
+// filterLastHeight removes ABCI requests from the last height if "commit" has not been called.
+func (g *GrammarChecker) filterLastHeight(reqs []*abci.Request) ([]*abci.Request, int) {
+	if len(reqs) == 0 {
+		return nil, 0
+	}
+	pos := len(reqs) - 1
+	r := reqs[pos]
+	cnt := 0
+	// Find the last commit.
+	for g.getRequestTerminal(r) != "commit" && pos > 0 {
+		pos--
+		r = reqs[pos]
+		cnt++
+	}
+	return reqs[:pos+1], cnt
+}
+
+// getRequestTerminal returns a value of a corresponding terminal in the ABCI grammar for a specific request.
+func (g *GrammarChecker) getRequestTerminal(req *abci.Request) string {
+	// req.String() produces an output like this "init_chain:<time:<seconds:-62135596800 > >"
+	// we take just the part before the ":" (init_chain, in previous example) for each request
+	s := req.String()
+	t := strings.Split(s, ":")[0]
+	return t
+}
+
+// getExecutionString returns a string of terminal symbols in parser readable format.
+func (g *GrammarChecker) getExecutionString(reqs []*abci.Request) string {
+	s := ""
+	for _, r := range reqs {
+		t := g.getRequestTerminal(r)
+		// We ensure to have one height per line for readability.
+		if t == "commit" {
+			s += t + "\n"
+		} else {
+			s += t + " "
+		}
+	}
+	return s
+}
+
+// Verify verifies whether a list of request satisfy ABCI grammar.
+func (g *GrammarChecker) Verify(reqs []*abci.Request, isCleanStart bool) (bool, error) {
+	if len(reqs) == 0 {
+		return false, fmt.Errorf("execution with no ABCI calls.")
+	}
+	r := g.filterRequests(reqs)
+	// Check if the execution is incomplete.
+	if len(r) == 0 {
+		return true, nil
+	}
+	var errors []*Error
+	execution := g.getExecutionString(r)
+	if isCleanStart {
+		errors = g.verifyCleanStart(execution)
+	} else {
+		errors = g.verifyRecovery(execution)
+	}
+	if errors == nil {
+		return true, nil
+	}
+	return false, fmt.Errorf("%v\nFull execution:\n%v", g.combineErrors(errors, g.cfg.NumberOfErrorsToShow), g.addHeightNumbersToTheExecution(execution))
+}
+
+// verifyCleanStart verifies if a specific execution is a valid clean-start execution.
+func (g *GrammarChecker) verifyCleanStart(execution string) []*Error {
```

Comment:
This function and the following are almost identical, IIUC, only differences are:
* the lexer you create
* the parser you call
* part of the error message

Can we factor out these two functions into an internal one accepting a lexer creation functor, a `Parse` functor, and a string "clean-start"/"recovery" to add to the error message?

Reply:
After exploring some solutions here, we've decided to leave this code as it is... as the alternatives end up adding unjustified complexity